rtcp/src/transport_feedbacks/transport_layer_nack/mod.rs

```diff
@@ -27,40 +27,69 @@ pub struct NackPair {
 pub type RangeFn =
     Box<dyn (Fn(u16) -> Pin<Box<dyn Future<Output = bool> + Send + 'static>>) + Send + Sync>;
 
-impl NackPair {
-    /// PacketList returns a list of Nack'd packets that's referenced by a NackPair
-    pub fn packet_list(&self) -> Vec<u16> {
-        let mut out = vec![self.packet_id];
+pub struct NackIterator {
+    packet_id: u16,
+    bitfield: PacketBitmap,
+    index: Option<u16>,
+}
+
+impl Iterator for NackIterator {
+    type Item = u16;
 
-        let mut b = self.lost_packets;
-        let mut i = 0;
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.index {
+            None => {
+                self.index = Some(0);
 
-        while b != 0 {
-            if (b & (1 << i)) != 0 {
-                b &= !(1 << i);
-                out.push(self.packet_id + i + 1);
+                Some(self.packet_id)
             }
-            i += 1;
+            Some(id) if self.bitfield != 0 => {
+                let mut i = id;
+
+                while self.bitfield != 0 {
+                    if (self.bitfield & (1 << i)) != 0 {
+                        self.bitfield &= !(1 << i);
+                        self.index = Some(i + 1);
+
+                        return Some(self.packet_id.wrapping_add(i + 1));
+                    }
+
+                    i += 1;
+                }
+
+                None
+            }
+            _ => None,
         }
-        out
+    }
+}
+
+impl NackPair {
+    /// PacketList returns a list of Nack'd packets that's referenced by a NackPair
+    pub fn packet_list(&self) -> Vec<u16> {
+        self.into_iter().collect()
     }
 
     pub async fn range(&self, f: RangeFn) {
-        if !f(self.packet_id).await {
-            return;
-        }
-
-        let mut b = self.lost_packets;
-        let mut i = 0u16;
-        while b != 0 {
-            if (b & (1 << i)) != 0 {
-                b &= u16::MAX ^ (1 << i);
-                let (packet_id, _) = self.packet_id.overflowing_add(i + 1);
-                if !f(packet_id).await {
-                    return;
-                }
+        for packet_id in self.into_iter() {
+            if !f(packet_id).await {
+                return;
             }
-            i += 1;
+        }
+    }
+}
+
+/// Create an iterator over all the packet sequence numbers expressed by this NACK pair.
+impl IntoIterator for NackPair {
+    type Item = u16;
+
+    type IntoIter = NackIterator;
+
+    fn into_iter(self) -> Self::IntoIter {
+        NackIterator {
+            packet_id: self.packet_id,
+            bitfield: self.lost_packets,
+            index: None,
```

Comment:
Wouldn't it be better to have `0` here? And increase by 1 for each iteration. It looks like this currently can oscillate, so you get:

None => 0 => None => 0

I.e. `iter.next()` would produce values again if you just "continue" after it produced the first `None`.

Reply:
`None`is different from `0`. 

For example, with the above values you used:

```
packet_id: 123
bitfield: `00000000 00000101`
```

The first call to `next` should produce `123` and also set `index` to `Some(0)` indicating that the next call to `next` should return a value based on the bitfield if any.

However, since each call to next that moves forward modifies `bitfield` I think we could get by with a `has_yielded_packet_id: bool` and always start the indexing from `0`, slightly less efficient but clearer.

Reply:
This is said alternative approach

```diff
diff --git a/rtcp/src/transport_feedbacks/transport_layer_nack/mod.rs b/rtcp/src/transport_feedbacks/transport_layer_nack/mod.rs
index 8761a433..43db9e9f 100644
--- a/rtcp/src/transport_feedbacks/transport_layer_nack/mod.rs
+++ b/rtcp/src/transport_feedbacks/transport_layer_nack/mod.rs
@@ -30,36 +30,30 @@ pub type RangeFn =
 pub struct NackIterator {
     packet_id: u16,
     bitfield: PacketBitmap,
-    index: Option<u16>,
+    has_yielded_packet_id: bool,
 }
 
 impl Iterator for NackIterator {
     type Item = u16;
 
     fn next(&mut self) -> Option<Self::Item> {
-        match self.index {
-            None => {
-                self.index = Some(0);
+        if !self.has_yielded_packet_id {
+            self.has_yielded_packet_id = true;
 
-                Some(self.packet_id)
-            }
-            Some(id) if self.bitfield != 0 => {
-                let mut i = id;
-
-                while self.bitfield != 0 {
-                    if (self.bitfield & (1 << i)) != 0 {
-                        self.bitfield &= !(1 << i);
-                        self.index = Some(i + 1);
+            Some(self.packet_id)
+        } else {
+            let mut i = 0;
 
-                        return Some(self.packet_id.wrapping_add(i + 1));
-                    }
+            while self.bitfield != 0 {
+                if (self.bitfield & (1 << i)) != 0 {
+                    self.bitfield &= !(1 << i);
 
-                    i += 1;
+                    return Some(self.packet_id.wrapping_add(i + 1));
                 }
 
-                None
+                i += 1;
             }
-            _ => None,
+            None
         }
     }
 }
@@ -89,7 +83,7 @@ impl IntoIterator for NackPair {
         NackIterator {
             packet_id: self.packet_id,
             bitfield: self.lost_packets,
-            index: None,
+            has_yielded_packet_id: false,
         }
     }
 }
diff --git a/rtcp/src/transport_feedbacks/transport_layer_nack/transport_layer_nack_test.rs b/rtcp/src/transport_feedbacks/transport_layer_nack/transport_layer_nack_test.rs
index e83f1f7c..e8c6b76f 100644
--- a/rtcp/src/transport_feedbacks/transport_layer_nack/transport_layer_nack_test.rs
+++ b/rtcp/src/transport_feedbacks/transport_layer_nack/transport_layer_nack_test.rs
@@ -199,6 +199,15 @@ fn test_nack_pair() {
             lost_packets: 0b0000_0111,
         },
     );
+
+    // Gap
+    test_nack(
+        vec![123, 124, 126],
+        NackPair {
+            packet_id: 123,
+            lost_packets: 0b0000_0101,
+        },
+    );
 }
 
 #[tokio::test]
```

Reply:
Let me know if you prefer this variant

Reply:
That is better, yes.

Reply:
Done